### PR TITLE
A possible finish for the jsx3 (lib) upgrade

### PR DESCRIPTION
**Pull request by hew**
**Pull Request Labels**

- [x] feature
- [x] blocking
- [ ] docs

I think this finished the jsx3 upgrade. I tried to approximately follow the style of how it was started. 

I haven't touched the examples files at all, yet.

### CODE REVIEW

**Comment by hew** [comment]
Note: this isn't blocking the core lib, just the examples compilation.

**Comment by Krankypanky** [comment]
Hey,
@hew I tried your repo and found out that the ApolloProvider is still not working.

I had to do a small change in the `ApolloProvider.re` which are:

```
open ApolloClient;

[@react.component]
[@bs.module "react-apollo"]
external make: (~client: generatedApolloClient, ~children: React.element) => React.element = "ApolloProvider";
```

so i can use it in my `index.re` like that:
```
ReactDOMRe.renderToElementWithId(
  <ReasonApollo.Provider client=Client.instance> <App /> </ReasonApollo.Provider>,
  "root",
);

```
Now it is working for me, or do i do something wrong?

**Comment by hew** [comment]
@Krankypanky 

I actually don't use the <Provider> component in reason land at work - which I why I completely forgot about it 😅. So yeah I'm assuming that `ApolloProvider.re` probably needs that child props change that you have done. I will push that onto the PR. Thanks!

**Comment by idkjs** [comment]
@hew Thanks for doing this work, brother. I wish I could have done it myself. 

I had to change `ApolloProvider.re` to this to get it to go.

```
open ApolloClient;

[@react.component] [@bs.module "react-apollo"]
external make: (~client: generatedApolloClient,~children:'a) => React.element = "ApolloProvider";
```

File: src/graphql-types/ReasonApolloMutation.re [hunk 1]
```diff
@@ -100,26 +100,84 @@ module Make = (Config: Config) => {
...
+        ~errorPolicy: option(errorPolicy)=?,
+        ~ssr: option(ssr)=?,
+        ~displayName: option(displayName)=?,
+        ~skip: option(skip)=?,
+        ~onCompleted: option(onCompleted)=?,
+        ~onError: option(onError)=?,
+        ~partialRefetch: option(partialRefetch)=?,
+        ~delay: option(delay)=?,
+        ~context: option(context)=?,
+        ~children: ('a, 'b) => React.element
```
Comment (hew) [inline src/graphql-types/ReasonApolloMutation.re]: This remains untyped. If anyone has any ideas here, that would be nice 😀

**Comment by hew** [comment]
@idkjs No problem! Yeah looks like you and Kranky came to the same conclusion. Adding that change now.

**Comment by hew** [comment]
Regarding the examples folder, I'm not totally sure what the overall direction there is. I see some unfinished hooks stuff. Maybe one of the library authors can chime in here.

**Comment by idkjs** [comment]
@hew about half way there on the swapi example.

**Comment by hew** [comment]
Looking through this repo's code, and [stuff like this](https://github.com/trojanowski/react-apollo-hooks/blob/master/src/useQuery.ts), I def got some ideas for a hooks-based API 😎

**Comment by hew** [comment]
Sorry - misclicked.

**Comment by idkjs** [comment]
@hew i was up in that repo earlier today while waiting on you to drop this pull request. Nice stuff in there, @trojanowski.

**Comment by Krankypanky** [comment]
@hew , @idkjs Im currently sitting for the query components, because my compiler is still screaming at me. I think i will be finished today or tomorrow, so i can share my code here, is that fine or is anybody else doing it right now?

**Comment by Gregoirevda** [comment]
@Krankypanky yes please, push it here if you started from this branch. Otherwise make a PR and I'll take your changes into account

Thanks a lot @hew to continue this! 👍 I didn't pushed everything on that branch, but I'll merge it all together. 

I'll release this asap as alpha release when finished and wait some time until enough feedback to deploy on latest.

**Comment by idkjs** [comment]
Swapi updated except for wto subscription components here: https://github.com/idkjs/reason-apollo/tree/remotes/origin/feature/jsx3/examples/swapi

Screwed up the git, need to walk away.

**Comment by hew** [comment]
@Gregoirevda 

> I'll release this asap as alpha 

Sounds good 🚀

**Comment by ulrikstrid** [comment]
@Gregoirevda and @hew it might make sense to keep both implementations for a while to allow for a smoother transition for codebases that already use this heavily

**Comment by Krankypanky** [comment]
Im currently fighting with this block:

```
[@react.component]
  let make =
      (
        ~variables: option(Js.Json.t)=?,
        ~pollInterval: option(int)=?,
        ~notifyOnNetworkStatusChange: option(bool)=?,
        ~fetchPolicy: option(string)=?,
        ~errorPolicy: option(string)=?,
        ~ssr: option(bool)=?,
        ~displayName: option(string)=?,
        ~skip: option(bool)=?,
        ~onCompleted: option(Js.Nullable.t(Js.Json.t) => unit)=?,
        ~onError: option(apolloError => unit)=?,
        ~partialRefetch: option(bool)=?,
        ~delay: option(bool)=?,
        ~context: option(Js.Json.t)=?,
        ~children: renderPropObj => React.element,
      ) =>
    ReasonReact.wrapJsForReason(
      ~reactClass=queryComponent,
      ~props=
        Js.Nullable.{
          "query": graphqlQueryAST,
          "variables": variables |> fromOption,
          "pollInterval": pollInterval |> fromOption,
          "notifyOnNetworkStatusChange":
            notifyOnNetworkStatusChange |> fromOption,
          "fetchPolicy": fetchPolicy |> fromOption,
          "errorPolicy": errorPolicy |> fromOption,
          "ssr": ssr |> fromOption,
          "displayName": displayName |> fromOption,
          "skip": skip |> fromOption,
          "onCompleted": onCompleted |> fromOption,
          "onError": onError |> fromOption,
          "partialRefetch": partialRefetch |> fromOption,
          "delay": delay |> fromOption,
          "context": context |> fromOption,
        },
      apolloData =>
      apolloData |> convertJsInputToReason |> children
    );
```

I have no clue how to wrap that part into the new syntax because we are not creating a component directly, we are creating a component with a function.... its a little bit confusing, any ideas?

this could look like this: 
```
[@react.component]
[@bs.module "react-apollo"]
external make: (....props) => React.element = "Query"; 
```

**Comment by MoOx** [comment]
>  it might make sense to keep both implementations for a while to allow for a smoother transition for codebases that already use this heavily

That's what we are going to do with bs-react-native:

- bs-react-native won't move
- bs-react-native-jsx3-compat
- reason-react-native (total rewrite, simpler interface more close to react-native)

Maybe multiple packages could help here.

**Comment by fakenickels** [comment]
Thanks for your help in the Reason community @hew!

**Comment by hew** [comment]
You're very welcome. Thanks for the fantastic community!

**Comment by elnygren** [comment]
btw. `ApolloConsumer` is also missing v3 support after this PR
(imo it's totally fine to handle that in a another PR to get this merged faster)

**Comment by idkjs** [comment]
@elnygren do you have a working example of the ApolloConsumer module that I might test against?

**Comment by Gregoirevda** [comment]
I finished https://github.com/apollographql/reason-apollo/tree/feature/jsx3 branch upgrade.
 swapi examples are converted too.

You can grab reason-apollo with JSX 3 support on npm under alpha release (v 0.16.0)
`yarn upgrade reason-apollo@alpha`

Please tell me if you encounter issues

**Comment by idkjs** [comment]
Serious work @Gregoirevda @hew and everyone else. Thank you.

**Pull request closed without merging**